### PR TITLE
AxisAlignedBB is now immutable

### LINKED ISSUE #88: Immutable AxisAlignedBB, Vector3 etc
No one expects the manipulation methods of AxisAlignedBB to alter the AABB directly. Allowing this is also a headache.

We already don't use Vector3 as mutable since PM 4.0.

In PHP 8.1, it's possible to use `readonly` to make all of the properties immutable. In 8.2, we'll be able to make whole classes readonly, but that also prevents inheritance by non-readonly classes.

### PR DESCRIPTION
**Pull request by DebHugo**
Like #90 but for AABB, fix #88

### CODE REVIEW

**Comment by dktapps** [comment]
I'd prefer if we kept the `*Copy` versions. That way, only the people using the mutable functions will get their code broken.

**Comment by DebHugo** [comment]
I said the same thing to myself but i think it's better to keep logic between Vector & AABB.
Vector is immutable and don't have a `copy` on the name of function return a new vector like `add` so i thinks it's better to do the same in AABB

**Comment by DebHugo** [comment]
Maybe add `@deprecated` for `*Copy` versions and remove them for pmmp 7 for exemple

**Comment by dktapps** [comment]
> I said the same thing to myself but i think it's better to keep logic between Vector & AABB. Vector is immutable and don't have a `copy` on the name of function return a new vector like `add` so i thinks it's better to do the same in AABB

Vector isn't relevant to this code. Totally different API. If anything, I'd've preferred `Vector` methods to have `Copy` suffixes to make it clear how they behave. But again, not relevant.

> Maybe add `@deprecated` for `*Copy` versions and remove them for pmmp 7 for exemple

I don't want to penalize correct code. There's no point doing a deprecation like that, it's just delaying the inevitable.
Either we keep them permanently, or we remove them now.

If we remove the `*Copy()` variations
- bad (mutable) code silently gets a behavioural BC break and possibly stops working
- good (immutable) code blows up and has to be refactored for no good reason

If we just remove the mutable variants and keep `*Copy()`
- bad (mutable) code blows up and has to be refactored (fine)
- good (immutable) code is unaffected

File: src/AxisAlignedBB.php [hunk 1]
```diff
@@ -181,44 +153,26 @@ public function contractedCopy(float $x, float $y, float $z) : AxisAlignedBB{
...
-	public function trim(Facing $face, float $distance) : AxisAlignedBB{
-		return $this->extend($face, -$distance);
-	}
-
-	/**
-	 * Returns a trimmed clone of this bounding box.
-	 * @see AxisAlignedBB::trim()
-	 */
-	public function trimmedCopy(Facing $face, float $distance) : AxisAlignedBB{
+	public function trimedCopy(Facing $face, float $distance) : AxisAlignedBB{
```
Comment (dktapps) [inline src/AxisAlignedBB.php]: typo

File: src/AxisAlignedBB.php [hunk 2]
```diff
@@ -229,43 +183,20 @@ public function trimmedCopy(Facing $face, float $distance) : AxisAlignedBB{
...
-	 * Returns a stretched copy of this bounding box.
-	 * @see AxisAlignedBB::stretch()
-	 */
 	public function stretchedCopy(Axis $axis, float $distance) : AxisAlignedBB{
-		return (clone $this)->stretch($axis, $distance);
+		return match($axis){
+			Axis::Y => new AxisAlignedBB($this->minX, $this->minY - $distance, $this->minZ, $this->maxX, $this->maxY + $distance, $this->maxZ),
+			Axis::Z => new AxisAlignedBB($this->minX, $this->minY, $this->minZ - $distance, $this->maxX, $this->maxY, $this->maxZ + $distance),
+			Axis::X => new AxisAlignedBB($this->minX - $distance, $this->minY, $this->minZ, $this->maxX + $distance, $this->maxY, $this->maxZ)
+		};
```
Comment (dktapps) [inline src/AxisAlignedBB.php]: The original code was more readable. Better to convert it to use local vars and then `return new AxisAlignedBB(...)` at the end.

File: src/AxisAlignedBB.php [hunk 3]
```diff
@@ -92,86 +92,58 @@ public function addCoord(float $x, float $y, float $z) : AxisAlignedBB{
 	 *
 	 * @return $this
```
Comment (dktapps) [inline src/AxisAlignedBB.php]: This is no longer correct

**Review by dktapps — CHANGES_REQUESTED** [review]
There are now many incorrect `@return $this` doc comments

File: src/AxisAlignedBB.php [hunk 4]
```diff
@@ -90,133 +90,96 @@ public function addCoord(float $x, float $y, float $z) : AxisAlignedBB{
...
-		$this->maxX -= $x;
-		$this->maxY -= $y;
-		$this->maxZ -= $z;
-
-		return $this;
-	}
-
-	/**
-	 * Returns a contracted clone of this AxisAlignedBB.
+	 * Returns an contracted clone of this AxisAlignedBB.
```
Comment (dktapps) [inline src/AxisAlignedBB.php]:
```suggestion
	 * Returns a contracted clone of this AxisAlignedBB.
```

**Review by dktapps — CHANGES_REQUESTED** [review]
Looking better, but I think the docs could be cleaned up

```
Outsets the bounds of this AxisAlignedBB by the specified X, Y and Z.

Returns an expanded clone of this AxisAlignedBB.
```
can be reworded like
```
Returns a copy of the AxisAlignedBB with bounds outset by the specified X, Y and Z.
```
etc

File: src/AxisAlignedBB.php [hunk 5]
```diff
@@ -88,183 +88,125 @@ public function addCoord(float $x, float $y, float $z) : AxisAlignedBB{
...
+			$this->minZ - $z,
+			$this->maxX + $x,
+			$this->maxY + $y,
+			$this->maxZ + $z
+		);
 	}
 
 	/**
-	 * Returns an offset clone of this AxisAlignedBB.
+	 * Returns a copy of the AxisAlignedBB with bounds offset by the specified X, Y and Z.
```
Comment (dktapps) [inline src/AxisAlignedBB.php]:
```suggestion
	 * Returns a copy of the AxisAlignedBB offset by the specified X, Y and Z.
```

File: src/AxisAlignedBB.php [hunk 6]
```diff
@@ -88,183 +88,125 @@ public function addCoord(float $x, float $y, float $z) : AxisAlignedBB{
...
+			$this->minZ + $z,
+			$this->maxX + $x,
+			$this->maxY + $y,
+			$this->maxZ + $z
+		);
 	}
 
 	/**
-	 * Returns an offset clone of this AxisAlignedBB.
+	 * Returns a copy of the AxisAlignedBB with bounds offset in the given direction by the specified distance.
```
Comment (dktapps) [inline src/AxisAlignedBB.php]:
```suggestion
	 * Returns a copy of the AxisAlignedBB offset in the given direction by the specified distance.
```

File: src/AxisAlignedBB.php [hunk 7]
```diff
@@ -88,183 +88,125 @@ public function addCoord(float $x, float $y, float $z) : AxisAlignedBB{
...
+		);
 	}
 
 	/**
-	 * Extends the AABB in the given direction.
-	 *
 	 * @param float $distance Negative values pull the face in, positive values push out.
 	 *
-	 * @return $this
+	 * Returns a copy of the AxisAlignedBB with bounds extended in the given direction.
```
Comment (dktapps) [inline src/AxisAlignedBB.php]: Summary should be placed above any @param or @return tags

Comment (dktapps) [reply src/AxisAlignedBB.php]: `Returns a copy of the AxisAlignedBB extended in the given direction.` would be more to the point also

File: src/AxisAlignedBB.php [hunk 8]
```diff
@@ -88,183 +88,125 @@ public function addCoord(float $x, float $y, float $z) : AxisAlignedBB{
...
-	 * @return $this
-	 */
-	public function trim(Facing $face, float $distance) : AxisAlignedBB{
-		return $this->extend($face, -$distance);
-	}
-
-	/**
-	 * Returns a trimmed clone of this bounding box.
-	 * @see AxisAlignedBB::trim()
+	 * Returns a copy of the AxisAlignedBB with bounds trimmed in the given direction.
```
Comment (dktapps) [inline src/AxisAlignedBB.php]: I don't think this is needed. Saying that it's the inverse of extend() is enough.

File: src/AxisAlignedBB.php [hunk 9]
```diff
@@ -88,183 +88,125 @@ public function addCoord(float $x, float $y, float $z) : AxisAlignedBB{
...
+	 * 
+	 * Inverse of extend().
```
Comment (dktapps) [inline src/AxisAlignedBB.php]:
```suggestion
	 * Inverse of extendedCopy().
```

File: src/AxisAlignedBB.php [hunk 10]
```diff
@@ -88,183 +88,125 @@ public function addCoord(float $x, float $y, float $z) : AxisAlignedBB{
...
 		return $this->extendedCopy($face, -$distance);
 	}
 
 	/**
-	 * Increases the dimension of the AABB along the given axis.
-	 *
 	 * @param float $distance Negative values reduce width, positive values increase width.
 	 *
-	 * @return $this
+	 * Returns a copy of the AxisAlignedBB with bounds stretched along the given axis.
```
Comment (dktapps) [inline src/AxisAlignedBB.php]: This should be above the @param tag, also `with bounds` seems superfluous

File: src/AxisAlignedBB.php [hunk 11]
```diff
@@ -88,183 +88,125 @@ public function addCoord(float $x, float $y, float $z) : AxisAlignedBB{
...
-	 */
-	public function squash(Axis $axis, float $distance) : AxisAlignedBB{
-		return $this->stretch($axis, -$distance);
+		return new AxisAlignedBB($minX, $minY, $minZ, $maxX, $maxY, $maxZ);
 	}
 
 	/**
-	 * Returns a squashed copy of this bounding box.
-	 * @see AxisAlignedBB::squash()
+	 * Returns a copy of the AxisAlignedBB with bounds squashed along the given axis.
```
Comment (dktapps) [inline src/AxisAlignedBB.php]: Unnecessary, saying it's inverse of stretch is sufficient

File: src/AxisAlignedBB.php [hunk 12]
```diff
@@ -88,183 +88,125 @@ public function addCoord(float $x, float $y, float $z) : AxisAlignedBB{
...
+	 * 
+	 * Inverse of stretch().
```
Comment (dktapps) [inline src/AxisAlignedBB.php]:
```suggestion
	 * Inverse of stretchedCopy().
```

**Review by dktapps — CHANGES_REQUESTED** [review]

**Review by dktapps — APPROVED** [review]

**Pull request merged**